Use childrenOfType<> more in HTMLTableElement

annevk opened the pull request:
#### 7cc92197012b63d00e2a9bfb97c775e7d0bef605
<pre>
Use childrenOfType&lt;&gt; more in HTMLTableElement
<a href="https://bugs.webkit.org/show_bug.cgi?id=124645">https://bugs.webkit.org/show_bug.cgi?id=124645</a>

Reviewed by Darin Adler.

* Source/WebCore/html/HTMLTableElement.cpp:
(WebCore::HTMLTableElement::tHead const):
(WebCore::HTMLTableElement::tFoot const):

Canonical link: <a href="https://commits.webkit.org/272797@main">https://commits.webkit.org/272797@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/44a6be3b2130aec823dbb16d56aae7fb636c7dac

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/33035 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/11808 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/34996 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/35675 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/29849 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/34006 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/14150 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/8978 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/29275 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/33510 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/9951 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/29551 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/8676 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/8830 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/29518 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/37008 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/30011 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/29916 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/34955 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/8968 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/6907 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/32810 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/10662 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7667 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/9557 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/9602 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/html/HTMLTableElement.cpp:
```diff
@@ -81,9 +81,9 @@ ExceptionOr<void> HTMLTableElement::setCaption(RefPtr<HTMLTableCaptionElement>&&
 RefPtr<HTMLTableSectionElement> HTMLTableElement::tHead() const
 {
-    for (RefPtr<Node> child = firstChild(); child; child = child->nextSibling()) {
+    for (Ref child : childrenOfType<HTMLTableSectionElement>(const_cast<HTMLTableElement&>(*this))) {
```
Seems like a `childrenOfType` bug that you need `const_cast` here.  Would be nice to fix it some day.

darinadler commented on Source/WebCore/html/HTMLTableElement.cpp:
```diff
@@ -81,9 +81,9 @@ ExceptionOr<void> HTMLTableElement::setCaption(RefPtr<HTMLTableCaptionElement>&&
+    for (Ref child : childrenOfType<HTMLTableSectionElement>(const_cast<HTMLTableElement&>(*this))) {
         if (child->hasTagName(theadTag))
-            return downcast<HTMLTableSectionElement>(child.get());
+            return child;
```
I wonder whether using `WTFMove` would be helpful, irrelevant, or harmful. We do want to hand off ownership, not churn the reference count, if possible.

annevk replied:
@cdumez could you help with this?

cdumez replied:
Like Darin, I'm uncertain here what's best, it would require testing. You might want to try but I have a vague recollection that gcc (and thus glib EWS) will complain about a redundant WTFMove() if you try to use it in such cases. If that's the case, it would seem to indicate that the WTFMove() is not needed.

annevk replied:
No compiler issue on macOS with adding WTFMove. Pushed it to the bots now.

cdumez replied:
gcc doesn't like it though:
```
In file included from /app/webkit/WebKitBuild/WPE/Release/WTF/Headers/wtf/FastMalloc.h:26,
                 from /app/webkit/Source/WebCore/config.h:47,
                 from /app/webkit/Source/WebCore/html/HTMLTableCaptionElement.cpp:25,
                 from /app/webkit/WebKitBuild/WPE/Release/WebCore/DerivedSources/unified-sources/UnifiedSource-950a39b6-14.cpp:1:
/app/webkit/Source/WebCore/html/HTMLTableElement.cpp: In member function ‘WTF::RefPtr<WebCore::HTMLTableSectionElement> WebCore::HTMLTableElement::tHead() const’:
/app/webkit/WebKitBuild/WPE/Release/WTF/Headers/wtf/StdLibExtras.h:675:58: error: redundant move in return statement [-Werror=redundant-move]
  675 | #define WTFMove(value) std::move<WTF::CheckMoveParameter>(value)
      |                        ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~
/app/webkit/Source/WebCore/html/HTMLTableElement.cpp:86:20: note: in expansion of macro ‘WTFMove’
   86 |             return WTFMove(child);
```

cdumez replied:
So my memory was correct and the original code was OK.

annevk replied:
Indeed. Will revert that change and land the prior patch then.

annevk committed:
Use childrenOfType<> more in HTMLTableElement
https://bugs.webkit.org/show_bug.cgi?id=124645

Reviewed by Darin Adler.

* Source/WebCore/html/HTMLTableElement.cpp:
(WebCore::HTMLTableElement::tHead const):
(WebCore::HTMLTableElement::tFoot const):

Canonical link: https://commits.webkit.org/272797@main
